### PR TITLE
[Feature] : Event visualiser enchancements

### PR DESCRIPTION
**Pull request by kshitij6325**
This PR contains following enhancements to event visualiser - 

- Showing clickstream connection status.
- Added search on event details page.
- Fix click on Event visualiser window

### Search on event details page.
Search on event details page

### Connection state change
 Connection state change

### CODE REVIEW

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasource.kt [hunk 1]
```diff
@@ -1,9 +1,12 @@
 package clickstream.eventvisualiser.ui.internal.data.datasource
 
 import clickstream.eventvisualiser.ui.internal.data.model.CSEvEvent
+import kotlinx.coroutines.flow.Flow
 
 internal interface CSEvDatasource {
 
+    val connectionFlow: Flow<Boolean>
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasource.kt:8]: how about just `isConnected`?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasource.kt [hunk 2]
```diff
@@ -15,6 +18,8 @@ internal interface CSEvDatasource {
         values: List<String> = listOf()
     ): List<String>
 
+    suspend fun getEventByNameAndId(eventId: String, eventName: String): CSEvEvent?
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasource.kt:21]: both arg should be true?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 3]
```diff
@@ -25,13 +32,16 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
...
+    override suspend fun addEvents(eventList: List<CSEvEvent>) {
+
     }
 
-    override suspend fun addEvents(eventList: List<CSEvEvent>) {
+    override val connectionFlow: Flow<Boolean>
+        get() = internalConnectionFlow.asStateFlow()
 
+    override fun startObserving() {
+        coroutineScope = CoroutineScope(SupervisorJob())
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:43]: should we cancel this scope in the stopObserving?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 4]
```diff
@@ -25,13 +32,16 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
         evObserver.removeObserver(eventCallback)
     }
 
-    override fun startObserving() {
-        coroutineScope = CoroutineScope(SupervisorJob())
-        evObserver.addObserver(eventCallback)
+    override suspend fun addEvents(eventList: List<CSEvEvent>) {
+
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:36]: shall we throw exception if someone called `addEvents`?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 5]
```diff
@@ -25,13 +32,16 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
...
     }
 
-    override suspend fun addEvents(eventList: List<CSEvEvent>) {
+    override val connectionFlow: Flow<Boolean>
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:39]: fields should be arrange before any functions

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 6]
```diff
@@ -67,43 +81,79 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
...
+
+    private fun handleEventData(csEvent: CSEventModel.Event) {
+        when (csEvent) {
+            is CSEventModel.Event.Instant -> addNewValueInCurrentList(
+                CSEvEvent(
+                    eventName = csEvent.eventName ?: "Empty event",
+                    eventId = csEvent.eventId,
+                    properties = csEvent.properties,
+                    state = CSEvState.ACKNOWLEDGED,
+                    timeStampInMillis = csEvent.timeStamp * 1000L,
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:109]: put 1000L as a constant?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 7]
```diff
@@ -67,43 +81,79 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
...
+                        // No-op
+                    }
+                }
+            }
+        }
+
+    private fun handleEventData(csEvent: CSEventModel.Event) {
+        when (csEvent) {
+            is CSEventModel.Event.Instant -> addNewValueInCurrentList(
+                CSEvEvent(
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:104]: the construction for CSEvEvent also duplicated with `CSEventModel.Event.Scheduled`, can we have an extract function for this?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt [hunk 8]
```diff
@@ -67,43 +81,79 @@ internal class CSEvDatasourceImpl private constructor(private val evObserver: CS
...
+                CSEvState.DISPATCHED
+            )
+        }
+    }
 
-                    is CSEventModel.Dispatched -> changeEventStatusInCurrentList(
-                        it,
-                        CSEvState.DISPATCHED
-                    )
+    private fun convertRawMessageToDisplayableString(message: String): String {
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:139]: do we have test for this function?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/repository/CSEvRepository.kt [hunk 9]
```diff
@@ -1,9 +1,12 @@
 package clickstream.eventvisualiser.ui.internal.data.repository
 
 import clickstream.eventvisualiser.ui.internal.data.model.CSEvEvent
+import kotlinx.coroutines.flow.Flow
 
 internal interface CSEvRepository {
 
+    val connectionFlow: Flow<Boolean>
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/repository/CSEvRepository.kt:8]: similar to https://github.com/gojek/clickstream-android/pull/25/files#r952239772

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/repository/CSEvRepositoryImpl.kt [hunk 10]
```diff
@@ -3,11 +3,17 @@ package clickstream.eventvisualiser.ui.internal.data.repository
 import clickstream.eventvisualiser.ui.internal.data.datasource.CSEvDatasource
 import clickstream.eventvisualiser.ui.internal.data.model.CSEvEvent
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.withContext
 
 
 internal class CSEvRepositoryImpl constructor(private val csEvDatasource: CSEvDatasource) :
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/repository/CSEvRepositoryImpl.kt:11]: how about to arrange to something like

```
internal class CSEvRepositoryImpl(
    private val csEvDatasource: CSEvDatasource
) : CSEvRepository {
```

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/ui/fragment/CSEvDetailListFragment.kt [hunk 11]
```diff
@@ -41,9 +46,39 @@ internal class CSEvDetailListFragment : CSEvBaseListFragment<FragmentCsEvDetailB
...
         super.onDestroy()
     }
 
+    private fun handleShowAllSwitch() {
+
+        binding?.swToggleAll?.setOnCheckedChangeListener { buttonView, isChecked ->
+            handleSwitchStateChange(isChecked)
+        }
+
+        viewLifecycleOwner.lifecycleScope.launchWhenStarted {
```
Comment (radityagumay) [inline clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/ui/fragment/CSEvDetailListFragment.kt:65]: why we need to use `viewLifecycleOwner.lifecycleScope.launchWhenStarted` twice?

File: clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt (hunk 3 above)
Comment (kshitij6325) [reply clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt:43]: We are cancelling it in `stopObserving` [here](https://github.com/gojek/clickstream-android/blob/ee102fb7091d08681205658f7fccdd4b587055e8/clickstream-event-visualiser-ui/src/main/java/clickstream/eventvisualiser/ui/internal/data/datasource/CSEvDatasourceImpl.kt#L31).

**Review by AbhijeetMallick — DISMISSED** [review]

**Review by AbhijeetMallick — APPROVED** [review]

**Pull request merged**